### PR TITLE
feat(backend): allowing auth users to create presets

### PR DESCRIPTION
**Pull request by axel7083**
## Description

This pull request allows users to create/update/delete presets. 

> ℹ️ This PR is only adding the backend logic to ease review.

## Related issue

Partially fixes https://github.com/decompme/decomp.me/issues/648

## Tests

- [x] Unit tests has been added to ensure expected behaviour

### CODE REVIEW

File: backend/coreapp/migrations/0042_migrate_presets_to_db.py [hunk 1]
```diff
@@ -823,6 +823,7 @@ def create_presets(apps: Apps, schema_editor: BaseDatabaseSchemaEditor) -> None:
 class Migration(migrations.Migration):
     dependencies = [
         ("coreapp", "0041_remove_project_and_github_stuff"),
+        ("coreapp", "0054_presets_owner"),
```
Comment (ethteck) [inline backend/coreapp/migrations/0042_migrate_presets_to_db.py:826]: is this intentional? I'm not a migration expert, but I don't think an older migration should depend on a newer one

File: backend/coreapp/platforms.py [hunk 2]
```diff
@@ -48,10 +49,15 @@ def to_json(
...
-        if include_presets:
+        if include_presets:       
+            # Here we are filtering the presets based on the owner
+            # if owner is None it means the preset is a system one => we keep it
+            # if owner is equal to profile it means it has been created by the user => we keep it
+            # otherwise we do not use it
             ret["presets"] = [
                 PresetSerializer(p).data
                 for p in Preset.objects.filter(platform=self.id).order_by("name")
+                if p.owner is None or p.owner == profile
```
Comment (ethteck) [inline backend/coreapp/platforms.py:60]: Everyone should be able to see all presets. If I own project Banana and make a preset for it, I want everyone else on the site to be able to use that preset who's working on that  game. 

If there become too many presets to easily handle, we can make UI changes to make it easier to find the ones you want. but for now I think we don't need to filter like this. but I get your concern

File: backend/coreapp/views/compiler.py [hunk 3]
```diff
@@ -34,11 +35,12 @@ def compilers_json() -> Dict[str, Dict[str, object]]:
     def platforms_json(
         include_presets: bool = True,
         include_num_scratches: bool = False,
+        profile: Optional[Profile] = None,
     ) -> Dict[str, Dict[str, object]]:
         ret: Dict[str, Dict[str, object]] = {}
 
         for platform in compilers.available_platforms():
-            ret[platform.id] = platform.to_json(include_presets, include_num_scratches)
+            ret[platform.id] = platform.to_json(include_presets, include_num_scratches, profile)
```
Comment (ethteck) [inline backend/coreapp/views/compiler.py:43]: likewise, I don't think we need any of this plumbing. the presets visible on the site shouldn't change based on who is fetching the data

File: backend/coreapp/views/preset.py [hunk 4]
```diff
@@ -13,31 +15,48 @@
...
 
-class ReadOnly(BasePermission):
-    def has_permission(self, request: Any, view: Any) -> bool:
-        return request.method in SAFE_METHODS
+class IsOwnerOrReadOnly(BasePermission):
+    def has_object_permission(self, request: Any, view: Any, obj: Any) -> bool:
+        if request.method in SAFE_METHODS:
+            return True
+        if isinstance(obj, Preset):
+            return obj.owner == request.profile
```
Comment (ethteck) [inline backend/coreapp/views/preset.py:35]: `or if request.profile.is_staff()`

if my syntax was wrong, we do this for making sure staff can delete scratches, so you should be able to use the same pattern

File: backend/coreapp/tests/test_preset.py [hunk 5]
```diff
@@ -53,6 +72,54 @@ def test_create_preset_not_authenticated(self) -> None:
         except AssertionError:
             pass
 
+    def test_user_create_preset(self) -> None:
```
Comment (ethteck) [inline backend/coreapp/tests/test_preset.py:75]: these tests are great! could you add one to make sure anons can't make presets?

**Review by ethteck — CHANGES_REQUESTED** [review]
awesome

File: backend/coreapp/migrations/0042_migrate_presets_to_db.py (hunk 1 above)
Comment (axel7083) [reply backend/coreapp/migrations/0042_migrate_presets_to_db.py:826]: There is a problem with the migration, and I was not sure how to solve it...

The `0042_migrate_presets_to_db.py` is adding items to the Preset table, however they are using the latest `Preset` model class, which has the `owner` column, therefore leading to an error, when step 42 is triggered before step 54, which add the column...

I am not a migration expect either, and would appreciate some help on this particular issue!

File: backend/coreapp/views/preset.py (hunk 4 above)
Comment (axel7083) [reply backend/coreapp/views/preset.py:35]: This is inside the `IsOwnerOrReadonly` and the permissiosn are defined as followed

https://github.com/decompme/decomp.me/blob/9bc7f8a073c81582ddbcb1d03fc83e9ce9a4bed9/backend/coreapp/views/preset.py#L40

Therefore not requiring admin to be checked. I can add a unit test to ensure the expected behaviour if necessary

File: backend/coreapp/tests/test_preset.py (hunk 5 above)
Comment (axel7083) [reply backend/coreapp/tests/test_preset.py:75]: There is already one which was already defined, I kept it.

https://github.com/decompme/decomp.me/blob/eca8cd2229974f7482ae77c0edff4a4786890d03/backend/coreapp/tests/test_preset.py#L66-L73

File: backend/coreapp/views/compiler.py (hunk 3 above)
Comment (axel7083) [reply backend/coreapp/views/compiler.py:43]: I agree, and will be removing it when I will get some time!

File: backend/coreapp/platforms.py (hunk 2 above)
Comment (axel7083) [reply backend/coreapp/platforms.py:60]: Good to know, I will restore previous code then

File: backend/coreapp/migrations/0042_migrate_presets_to_db.py (hunk 1 above)
Comment (axel7083) [reply backend/coreapp/migrations/0042_migrate_presets_to_db.py:826]: This behaviour seems to have already been done in previous migration:

https://github.com/decompme/decomp.me/blob/6efceba4302d7064ec3d0614c5ed0bd12aed9424/backend/coreapp/migrations/0003_github_profile.py#L11

Comment (ethteck) [reply backend/coreapp/migrations/0042_migrate_presets_to_db.py:826]: I believe the way to handle this is the migration code shouldn't actually be using the real model classes, but I'm not 100% sure. we probably have screwed this up in the past. I'll see if I can take a look later today

File: backend/coreapp/views/preset.py (hunk 4 above)
Comment (ethteck) [reply backend/coreapp/views/preset.py:35]: ahhh missed that! thanks :)

File: backend/coreapp/views/compiler.py (hunk 3 above)
Comment (axel7083) [reply backend/coreapp/views/compiler.py]: Done ✅

File: backend/coreapp/migrations/0042_migrate_presets_to_db.py (hunk 1 above)
Comment (axel7083) [reply backend/coreapp/migrations/0042_migrate_presets_to_db.py:826]: Keep me updated

**Review by ethteck — APPROVED** [review]
other than the mypy issue, lgtm!

**Comment by axel7083** [comment]
@ethteck now mypy should be happy :)

**Pull request merged**